src/confirm.rs

```diff
@@ -1,48 +1,230 @@
+use helpers::CallResult;
 use {Transport, Error};
 
-pub fn wait_for_confirmations<'a, T, F, V>(transport: T, poll_interval: Duration, confirmations: usize, check: V)
-  -> Box<Future<Item = (), Error = Error> + 'a> where
-  T: 'a + Transport + Clone,
-  F: 'a + IntoFuture<Item = Option<U256>, Error = Error>,
-  V: 'a + Fn() -> F,
-{
-  let eth = EthFilter::new(transport.clone());
-  let result = eth.create_blocks_filter()
-    .and_then(move |filter| {
-      filter.stream(poll_interval)
-        .skip(confirmations as u64)
-        .and_then(move |_| check())
-        .filter_map(|o| o)
-        .and_then(move |confirmed_block_number| {
-          Eth::new(transport.clone()).block_number()
-            .map(move |last_block_number| confirmed_block_number.low_u64() + confirmations as u64 >= last_block_number.low_u64())
-        })
-        .filter(|confirmed| *confirmed)
-        .take(1)
-        .collect()
-        .map(|_| ())
-    });
-  Box::new(result)
-}
-
-pub fn send_transaction_with_confirmation<'a, T>(transport: T, tx: TransactionRequest, poll_interval: Duration, confirmations: usize) -> Box<Future<Item = TransactionReceipt, Error = Error> + 'a> where
-  T: 'a + Transport + Clone {
-  let eth = Eth::new(transport.clone());
-  let result = eth.send_transaction(tx)
-    .and_then(move |hash| {
-      wait_for_confirmations(transport.clone(), poll_interval, confirmations, move || {
-        eth.transaction_receipt(hash.clone()).map(|option| option.map(|receipt| receipt.block_number))
-      })
-      .and_then(move |_| {
-        let eth = Eth::new(transport.clone());
-        eth.transaction_receipt(hash).and_then(|option| option.ok_or(Error::Unreachable))
+pub trait ConfirmationCheck {
+  type Check: IntoFuture<Item = Option<U256>, Error = Error>;
+
+  fn check(&self) -> Self::Check;
+}
+
+enum WaitForConfirmationsState<F, O> {
+  WaitForNextBlock,
+  CheckConfirmation(F),
+  CompareConfirmations(u64, CallResult<U256, O>),
+}
+
+struct WaitForConfirmations<T, V, F> where T: Transport {
+  transport: T,
+  state: WaitForConfirmationsState<F, T::Out>,
+  filter_stream: Skip<FilterStream<T, H256>>,
+  confirmation_check: V,
+  confirmations: u64,
+}
+
+impl<T, V, F> Future for WaitForConfirmations<T, V, F::Future> where
+  T: Transport,
+  V: ConfirmationCheck<Check = F>,
+  F: IntoFuture<Item = Option<U256>, Error = Error> {
+
+  type Item = ();
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    loop {
+      let next_state = match self.state {
+        WaitForConfirmationsState::WaitForNextBlock => {
+          let _ = try_ready!(self.filter_stream.poll());
+          WaitForConfirmationsState::CheckConfirmation(self.confirmation_check.check().into_future())
+        },
+        WaitForConfirmationsState::CheckConfirmation(ref mut future) => match try_ready!(future.poll()) {
+          Some(confirmation_block_number) => {
+            let future = Eth::new(&self.transport).block_number();
```

Comment:
we can, but then we make redundant calls to get block number when `confirmation_check.check()` returns `None`